extensions/in-memory/assetindex-memory/src/test/java/org/eclipse/dataspaceconnector/metadata/memory/InMemoryAssetIndexTest.java

```diff
@@ -0,0 +1,163 @@
+        assertThat(assets).isEmpty();
+    }
+
+    @Test
+    void queryAssets_multipleFound() {
+        var testAsset1 = Asset.Builder.newInstance().id(UUID.randomUUID().toString()).name("foobar").version("1").build();
+        var testAsset2 = Asset.Builder.newInstance().id(UUID.randomUUID().toString()).name("barbaz").version("1").build();
+        var testAsset3 = Asset.Builder.newInstance().id(UUID.randomUUID().toString()).name("barbaz").version("1").build();
+        index.add(testAsset1, createDataAddress(testAsset1));
+        index.add(testAsset2, createDataAddress(testAsset2));
+        index.add(testAsset3, createDataAddress(testAsset3));
+        var assets = index.queryAssets(AssetSelectorExpression.Builder.newInstance()
+                .whenEquals("name", "barbaz")
+                .whenEquals("version", "1")
+                .build());
+        assertThat(assets).hasSize(2).containsExactlyInAnyOrder(testAsset2, testAsset3);
+    }
+
+    @Test
+    void queryAssets_noExpression_shouldReturnEmpty() {
+        var result = index.queryAssets(AssetSelectorExpression.Builder.newInstance().build());
+        assertThat(result).isEmpty();
+    }
+
+    @Test
+    void queryAssets_selectAll_shouldReturnAll() {
+        var testAsset1 = Asset.Builder.newInstance().id(UUID.randomUUID().toString()).name("barbaz").version("1").build();
+        index.add(testAsset1, createDataAddress(testAsset1));
+
+        var testAsset2 = Asset.Builder.newInstance().id(UUID.randomUUID().toString()).name("foobar").version("1").build();
+        index.add(testAsset2, createDataAddress(testAsset2));
+
+        assertThat(index.queryAssets(AssetSelectorExpression.SELECT_ALL)).containsExactlyInAnyOrder(testAsset1, testAsset2);
+    }
+
+    @Test
+    void findById() {
+        String id = UUID.randomUUID().toString();
+        var testAsset = Asset.Builder.newInstance().id(id).name("foobar").version("1").build();
+        index.add(testAsset, createDataAddress(testAsset));
+
+        assertThat(index.findById(id)).isNotNull().isEqualTo(testAsset);
+    }
+
+    @Test
+    void findById_notfound() {
+        String id = UUID.randomUUID().toString();
+        var testAsset = Asset.Builder.newInstance().id(id).name("foobar").version("1").build();
+        index.add(testAsset, createDataAddress(testAsset));
+
+        assertThat(index.findById("not-exist")).isNull();
+    }
+
+    @Test
+    void findById_multiple_raisesException() {
+        String id = UUID.randomUUID().toString();
+        var testAsset = Asset.Builder.newInstance().id(id).name("foobar").version("1").build();
+        var testAsset2 = Asset.Builder.newInstance().id(id).name("foobar").version("1").build();
+
+        index.add(testAsset, createDataAddress(testAsset));
+        index.add(testAsset2, createDataAddress(testAsset2)); //in an actual DB this would likely cause problems already
```

Comment:
This is an interesting comment. Maybe you can also explain why this leads to problems. So that future developers, that see these comment, understand the reasoning.

Reply:
Maybe would be better to get the error on the `add` call directly?

Reply:
i changed the in mem store to use a `Map`, so it becomes a non-issue.